fix: Filter for folders in cleanup old preview job

hammer065 opened the pull request:
* Resolves: #35936

## Summary
When running `OC\Preview\BackgroundCleanupJob`, the main iteration loop in `run()` expects a folder,
however, `getOldPreviewLocations()` currently does not filter by mimetype
and therefore can yield a non-folder entry which causes an Exception when constructing the Folder impl.
Filtering for `httpd/unix-directory`, as `getNewPreviewLocations()` already does, fixes this issue.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

kesselb commented:
Same as https://github.com/nextcloud/server/pull/46072?

hammer065 commented:
As the above mentioned PR seems to be staled, I rebased and refined this one.

It now also incorporates the second improvement made in 0c7eeec43f50632c19513697f5ceed056a6e8c69 for the job also ignoring non-numeric named directories.
However, 0c7eeec43f50632c19513697f5ceed056a6e8c69 did this inefficiently by needing to cast *all* `filecache` row IDs to a string, which then also results in needing to perform expensive string comparisons for the `JOIN`.
The same goal is still achieved by further filtering the small amount of entries for this appid of type directory to match against a trivial RegExp.
As an expression to match against a RegExp was not present yet, I implemented it for this PR.

Requesting a review @artonge @kesselb @artonge

solracsf commented:
Can you please explain why this PR is better than https://github.com/nextcloud/server/pull/46072 ?

hammer065 commented:
@solracsf Sure thing:
1. As I explained in https://github.com/nextcloud/server/pull/48581#issuecomment-2604447290, the changes in the other PR contain an majorly inefficient query change, which results in casting the numeric ID for *all* files/folders on the Nextcloud instance to a string to then be string-compared against a (to be expected) small list of files, fully bypassing any index, slowing down the periodic `BackgroundCleanupJob`.

2. This PR contains tests for the incorporated changes.

come-nc reviewed (approved):
This REGEX thing is a bit scary but I suppose if the tests pass on all DB we should be fine?

artonge reviewed (commented):
Is the regexp part actually needed? What kind of directory were selected when they should not be?

come-nc commented:
@hammer065 Thank you for your work on this and explanations.
What we were wondering when discussing this issue with @artonge is that the join on (int)name = fileid should already filter out non-numeric folder names. But there may be corner cases depending on DB servers, I remember that cast can behave quite weird in some situations.
Like '1.3' could be cast to 1 and not be filtered out on sqlite I think.

But that’s quite a niche cornercase, and the code after should survive. Worst case scenario I think it will delete a folder in the preview folder which should not be there in the first place.

So all in all I would vote for the simpler version of the PR without the regex usage. Maybe we can keep the regex implementation the queryBuilder though, it may prove useful some day.

artonge commented:
Thanks for the detailed summary @hammer065. On my side, I indeed don't see the point in introducing such logic for a hypothetical edge case.

> my approach prevents the costly cast of all file-IDs to a string by just filtering out all non-numeric folder names using the RegExp

I would especially be against, as "all non-numeric folder names" would probably be only one.

artonge commented:
> Maybe we can keep the regex implementation the queryBuilder though, it may prove useful some day.

Not sure about the performance of such operator. Unless it is fast, having it in by default might lead to being it use without knowing better and causing performance issues. If we don't have a use case, let's not keep it.

hammer065 commented:
Please let me know what the end position is after your (internal) discussions, so I can adapt the PR accordingly.

My personal stances are to both *not* support arbitrary *folders*, just files, in the previews directory and *not* to keep the RegEx code.

come-nc commented:
> Please let me know what the end position is after your (internal) discussions, so I can adapt the PR accordingly.
> 
> My personal stances are to both _not_ support arbitrary _folders_, just files, in the previews directory and _not_ to keep the RegEx code.

We agree, please remove regex related code and keep the changes minimal, it will also be easier to backport.

hammer065 committed:
fix(preview): Filter for folders in cleanup old preview job

Fixes #35936.
When running `OC\Preview\BackgroundCleanupJob`, the main iteration loop
in `run()` expects a folder, however, `getOldPreviewLocations()`
currently does not filter by mimetype and therefore can yield a
non-folder entry which causes an Exception when constructing the Folder
impl.
Filtering for `httpd/unix-directory`, as `getNewPreviewLocations()`
already does, fixes this issue.

Signed-off-by: Dario Mehlich <d.mehlich@gmail.com>

hammer065 commented:
> We agree, please remove regex related code and keep the changes minimal, it will also be easier to backport.

@come-nc Done, you can review this now (finally) final commit :D

artonge commented on lib/private/Preview/BackgroundCleanupJob.php:
```diff
@@ -63,13 +63,13 @@ private function getOldPreviewLocations(): \Iterator {
-				$qb->expr()->eq('a.parent', $qb->createNamedParameter($this->previewFolder->getId()))
-			)->andWhere(
-				$qb->expr()->like('a.name', $qb->createNamedParameter('__%'))
+				$qb->expr()->andX(
+					$qb->expr()->isNull('b.fileid'),
+					$qb->expr()->eq('a.storage', $qb->createNamedParameter($this->previewFolder->getStorageId())),
+					$qb->expr()->eq('a.parent', $qb->createNamedParameter($this->previewFolder->getId())),
+					$qb->expr()->like('a.name', $qb->createNamedParameter('__%')),
+					$qb->expr()->eq('a.mimetype', $qb->createNamedParameter($this->mimeTypeLoader->getId('httpd/unix-directory')))
+				)
```
I would have kept the `andWhere` syntax to reduce the amount of changes, but good anyway

hammer065 replied:
I have done it as, in my opinion, the `->andWhere()` syntax is more verbose and less readable than the `->andX()` syntax. It additionally makes this consistent to the nearly identical query in L133-145.

artonge replied:
Consistency and low change count was my concern. If it is consistent with the sibling query, then perfect :+1:

come-nc commented:
/backport to stable31

come-nc commented:
/backport to stable30

come-nc commented:
/backport to stable29

come-nc commented:
@hammer065 Thanks a lot for your patience and team work here!

Would you be interested in a Guest account on our Nextcloud Talk instance where we host a dev chat?

hammer065 commented:
> Would you be interested in a Guest account on our Nextcloud Talk instance where we host a dev chat?
@come-nc Sure thing, would definitely speed up the discussion process for any potential future PRs :D
Can this be linked/Is this linkable to my account on my own Nextcloud instance or would it be a separate account all together?

come-nc commented:
> > Would you be interested in a Guest account on our Nextcloud Talk instance where we host a dev chat?
> > @come-nc 
>
> Sure thing, would definitely speed up the discussion process for any potential future PRs :D
> Can this be linked/Is this linkable to my account on my own Nextcloud instance or would it be a separate account all together?

I invited you on the email used for the commit sign-off.
It’s a guest account using the guests app, it’s not linkable to a federated account. I think we did have federation enabled for Talk on our instance, but I’m not sure whether it’s enabled for community chat yet.
